File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(string) {
    if(string.length<=1) return string.length;
    let left = 0, longest=0;
    const seenChars = {};
    
    for(let right = 0; right<string.length; right++){
        let currentChar = string[right];
        let previouslySeenChar = seenChars[currentChar];
        if(previouslySeenChar >= left) {
            left = previouslySeenChar + 1;
        }
        
        seenChars[currentChar] = right;
        longest = Math.max(longest, right-left+1)
    }
    
    return longest;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if(s.Length == 0) return 0;
        
        int maxLength = 1;
        
        for(int i = 0; i < s.Length; i++) {
            HashSet<char> hashset = new HashSet<char>();
            hashset.Add(s[i]);
            int currLength = 1;
            currLength = dfs(i+1, hashset, s, currLength);
            maxLength = Math.Max(maxLength, currLength);
        }
        
        return maxLength;
    }
    
    public int dfs(int index, HashSet<char> hashset, string s, int currLength) {
        if(index == s.Length || hashset.Contains(s[index])) {
            return currLength;
        }
        currLength++;
        hashset.Add(s[index]);
        return dfs(index+1, hashset, s, currLength);
    }
}